services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/util/DeviceRegistryUtils.java

```diff
@@ -160,22 +159,26 @@ public static String getUniqueIdentifier() {
      * @param span The active OpenTracing span for this operation. It is not to be closed in this method! An
      *             implementation should log (error) events on this span and it may set tags and use this span 
      *             as the parent for any spans created in this method.
-     * @return the certificate of the device to be provisioned.
+     * @return A future indicating the outcome of the operation. If the operation succeeds, the
+     *         retrieved certificate is returned. Else {@link Optional#empty()} is returned.
      * @throws NullPointerException if any of the parameters except span is {@code null}.
-     * @throws ClientErrorException if error getting the certificate from the given client context.
      */
-    public static X509Certificate getCertificateFromClientContext(
+    public static Future<Optional<X509Certificate>> getCertificateFromClientContext(
             final String tenantId,
             final String authId,
             final JsonObject clientContext,
-            final Span span) throws ClientErrorException {
+            final Span span) {
 
         Objects.requireNonNull(tenantId);
         Objects.requireNonNull(authId);
         Objects.requireNonNull(clientContext);
 
         try {
             final byte[] bytes = clientContext.getBinary(CredentialsConstants.FIELD_CLIENT_CERT);
+            if (bytes == null) {
+                throw new IllegalArgumentException(
+                        String.format("The client context doesn't contain a certificate for tenant [%s]", tenantId));
```

Comment:
this just means that the adapter doesn't support auto provisioning, right? So, FMPOV this should result in
```
return Optional.empty();
```

Reply:
I am not deep into this auto-provisioning topic. As per my understanding, we are checking first if  auto-provisioning is enabled using `isAutoProvisioningEnabled(...)`. If it is enabled, then  `getCertificateFromClientContext(...)` is being invoked. Since it is already enabled, I thought  that the client context should return a certificate and hence I made it so.
I will replace it with ` return Optional.empty();`